_posts/2020_04_29_encrypted_keras_with_syft/encrypted_keras_with_syft.Rmd

```diff
@@ -0,0 +1,413 @@
+
+#### Prelude: Train a simple model on MNIST
+
+```{r}
+# create_model.R
+
+library(tensorflow)
+library(keras)
+
+mnist <- dataset_mnist()
+mnist$train$x <- mnist$train$x/255
+mnist$test$x <- mnist$test$x/255
+
+dim(mnist$train$x) <- c(dim(mnist$train$x), 1)
+dim(mnist$test$x) <- c(dim(mnist$test$x), 1)
+
+input_shape <- c(28, 28, 1)
+
+model <- keras_model_sequential() %>%
+  layer_conv_2d(filters = 16, kernel_size = c(3, 3), input_shape = input_shape) %>%
+  layer_average_pooling_2d(pool_size = c(2, 2)) %>%
+  layer_activation("relu") %>%
+  layer_conv_2d(filters = 32, kernel_size = c(3, 3)) %>%
+  layer_average_pooling_2d(pool_size = c(2, 2)) %>%
+  layer_activation("relu") %>%
+  layer_conv_2d(filters = 64, kernel_size = c(3, 3)) %>%
+  layer_average_pooling_2d(pool_size = c(2, 2)) %>%
+  layer_activation("relu") %>%
+  layer_flatten() %>%
+  layer_dense(units = 10, activation = "linear")
+  
+
+model %>% compile(
+  loss = "sparse_categorical_crossentropy",
+  optimizer = "adam",
+  metrics = "accuracy"
+)
+
+model %>% fit(
+    x = mnist$train$x,
+    y = mnist$train$y,
+    epochs = 1,
+    validation_split = 0.3,
+    verbose = 2
+)
+
+model$save(filepath = "model.hdf5")
+```
+
+#### Set up cluster and serve model
+
+The easiest way to get all required packages is to install the ensemble OpenMined put together for their [Udacity
+Course](https://www.udacity.com/course/secure-and-private-ai--ud185) that introduces federated learning and differential
+privacy with PySyft. This will install TensorFlow 1.15 and TensorFlow Encrypted, amongst others.
+
+The following lines of code should all be put together in a single file. I found it practical to "source" this script from an
+R process running in a console tab.
+
+To begin, we again define the model, two things being different now. First, for technical reasons, we need to pass in
+`batch_input_shape` instead of `input_shape`. Second, the final layer is "missing" the softmax activation. This is not an
+oversight -- SMPC `softmax` has not been implemented yet. (Depending on when you read this, that statement may no longer be
+true.) Were we training this model in *secret sharing* mode, this would of course be a problem; for classification though, all
+we care about is the maximum score.
+
+After model definition, we load the actual weights from the model we trained in the previous step. Then, the action begins. We
+create an ensemble of TFE workers that together run a distributed TensorFlow cluster. The model is *secret* *shared* with the
+workers, that is, model weights are split up into shares that, each inspected alone, are unusable. Finally, the model is
+*served*, i.e., made available to clients requesting predictions.
+
+How can a Keras model be *shared* and *served*? These are not methods provided by Keras itself. The magic is due to Syft
+*hooking* into Keras, extending the `model` object: cf. `hook <- sy$KerasHook(tf$keras)` in line 10.
```

Comment:
Are the lines of code visibly numbered in the code chunks of the blog? If not, I'd just reference the code, and not mention the line number.

Reply:
oh, you're right! The code has line numbers in the visual editor :-)